src/abstractdataframe/nest.jl

```diff
@@ -0,0 +1,390 @@
+           promote::Bool=true,
+           makeunique::Bool=false, flatten::Bool=true)
+
+Unnest one or more columns `src` into multiple columns. The newly created
+columns are stored at the end of the data frame (and the `src` column is
+dropped).
+
+Each `src` column must contain a `NamedTuple`, a `DataFrameRow`, a
+`Tables.AbstractRow`, or Tables.jl table.
+
+`cols` (default `:setequal`) and `promote` (default `true`) keyword arguments
+have the same meaning as in [`push!`](@ref).
+
+If `makeunique=false` (the default) produced column names must be unique.
+If `makeunique=true` then duplicate column names will be suffixed with `_i`
+(`i` starting at `1` for the first duplicate).
+
+If `flatten=true` (the default) then newly created columns are flattened
+using [`flatten`](@ref) with `scalar=Missing` keyword argument.
+
+TODO: metadata
+
+"""
+function unnest(df::AbstractDataFrame, src::ColumnIndex...;
+                cols::Union{Symbol, AbstractVector{Symbol},
+                            AbstractVector{<:AbstractString}}=:setequal,
+                promote::Bool=(cols in [:union, :subset]),
+                makeunique::Bool=false, flatten::Bool=true)
+    ref_df = select(df, Not(collect(Any, src)))
+    col_count = ncol(ref_df)
+    for idx in src
+        col = df[!, idx]
+        tmp_df = DataFrame()
+        for v in col
+            if v isa DataFrame # produce DataFrameRow
+                # if flatten=false make a copy to avoid aliases
+                v = DataFrame([n => [flatten ? c : copy(c)]
+                               for (n, c) in pairs(eachcol(v))],
+                              copycols=false) |> only
+            elseif Tables.istable(v) # produce NamedTuple
+                v = Tables.columntable(v)
+            end
+            push!(tmp_df, v, cols=cols, promote=promote)
+        end
+        hcat!(ref_df, tmp_df, makeunique=makeunique, copycols=false)
+    end
+    return if flatten
+        DataFrames.flatten(ref_df, col_count+1:ncol(ref_df), scalar=Missing)
+    else
+        ref_df
+    end
+end
+
+"""
+    flatten(df::AbstractDataFrame, cols; scalar::Type)
+
+When columns `cols` of data frame `df` have iterable elements that define
+`length` (for example a `Vector` of `Vector`s), return a `DataFrame` where each
+element of each `col` in `cols` is flattened, meaning the column corresponding
+to `col` becomes a longer vector where the original entries are concatenated.
+Elements of row `i` of `df` in columns other than `cols` will be repeated
+according to the length of `df[i, col]`. These lengths must therefore be the
+same for each `col` in `cols`, or else an error is raised. Note that these
+elements are not copied, and thus if they are mutable changing them in the
+returned `DataFrame` will affect `df`.
+
+`cols` can be any column selector ($COLUMNINDEX_STR; $MULTICOLUMNINDEX_STR).
+
+If `scalar` is passed then values that have this type in flattened columns
```

Comment:
Will I need to explicitly pass `scalar=Union{String,Missing}` for such columns to avoid splatting strings into chars?

Reply:
or e.g. `scalar=AbstractString` if you do not care about `missing`.

We could make `scalar=Union{Missing, AbstractString}` the default, but it would be breaking (as these are the most common cases people would care about). Still - maybe we could allow for it as it is only slightly breaking and it seems most people feel this behavior is preferred.

As a starting point I proposed a non-breaking design (i.e. doing what the function does now by default, but allowing a different behavior if needed).

Reply:
I think changing the default is better. I'm generally annoyed that strings are iterable and I'd rather not have to remember that they are.

Reply:
I agree. But the tension here is "better" vs "breaking". Therefore this needs a consensus in JuliaData community.

In general we have three options for `AbstractString` as default:
* expand (the current behavior)
* error (special case `AbstractString` in the code and throw an error if someone tries to expand it, unless it is included in `scalar`)
* treat as scalar by default (and broadcast if needed)

Reply:
So I propose to just keep the current behavior and add `scalar` kwarg (so keep the implementation I have proposed now).

I know that broadcast mechanics was previously proposed, but it is not correct either. The reason is:
```
julia> length(Base.Broadcast.broadcastable("xyz"))
1
```
so string would be seen as length-1 collection (and this is different from it being seen as a scalar).

Reply:
Yeah, it's a bit tricky. Maybe `length` isn't the right operation. We could check whether `ndims(Base.Broadcast.broadcastable(x)) == 0`, which is the closest equivalent of a scalar in the context of broadcasting. Not sure whether that's worth a breaking change.

Reply:
I agree - and for this reason I would just leave the things as they are - just add `scalar` kwarg that is non-breaking.